### PR TITLE
Fix disabled formfields trigger validation

### PR DESCRIPTION
**Pull request by gitvadim**
Hey,

Not sure if it's a bug or not, but I've discovered that FormFields' validate method is executed even if that field is disabled. This behavior prevents forms from submitting if you have disabled fields with validation on them.

P.S.

I've never done contributions on GitHub yet, so I hope I'm doing it right. I've read the contributing guide, but it hasn't made things easier to me :(

### CODE REVIEW

**Comment by gitvadim** [comment]
I don't know why, but the Google bot shows that I am yet to accept the CLA, but I've already done that.

File: packages/flutter/lib/src/widgets/form.dart [hunk 1]
```diff
@@ -183,17 +183,17 @@ class _FormScope extends InheritedWidget {
...
 ///
 /// Used by [FormField.onSaved].
-typedef void FormFieldSetter<T>(T newValue);
+typedef FormFieldSetter<T> = void Function(T newValue);
 
 /// Signature for building the widget representing the form field.
 ///
 /// Used by [FormField.builder].
-typedef Widget FormFieldBuilder<T>(FormFieldState<T> field);
+typedef FormFieldBuilder<T> = Widget Function(FormFieldState<T> field);
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/form.dart]: these seem like unrelated changes, best to keep those to a separate PR

File: packages/flutter/lib/src/widgets/form.dart [hunk 2]
```diff
@@ -256,6 +257,9 @@ class FormField<T> extends StatefulWidget {
   /// autovalidates, this value will be ignored.
   final bool autovalidate;
 
+  /// If true, indicates that the widget is enabled
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/form.dart]: please end sentences with a period. Also, maybe add a second paragraph explaining the implications (e.g. does this prevent interaction, prevent validation, etc)

File: packages/flutter/lib/src/widgets/form.dart [hunk 3]
```diff
@@ -279,7 +283,7 @@ class FormFieldState<T> extends State<FormField<T>> {
 
   /// Calls the [FormField]'s onSaved method with the current value.
   void save() {
-    if (widget.onSaved != null)
+    if (widget.onSaved != null && widget.enabled)
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/form.dart]: we should definitely document that this does nothing if enabled is false, but is that even what we want? If someone calls save() on a disabled field, should that work? If not, should it assert?

File: packages/flutter/lib/src/widgets/form.dart [hunk 4]
```diff
@@ -301,7 +305,9 @@ class FormFieldState<T> extends State<FormField<T>> {
   }
 
   bool _validate() {
-    if (widget.validator != null)
+    if (!widget.enabled)
+      _errorText = null;
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/form.dart]: it seems like if it was disabled, we should do nothing, rather than clear the previous state.
on the other hand, if _validate() is being called, even if disabled, maybe we should validate regardless? and just not call it if we're disabled?

Comment (Hixie) [reply packages/flutter/lib/src/widgets/form.dart]: Looking at the code, it seems like it might be better to just not autovalidate when disabled, but otherwise to continue working as now. That way if someone _wants_ to call validate() on a disabled field, it works ok.

**Comment by Hixie** [comment]
I'm not sure I understand this patch. It seems to _add_ the ability to make a text field disabled, but it doesn't seem to actually disable the text field in that case, just disabled validation and saving? Can you elaborate more on your PR? Thanks.

File: packages/flutter/lib/src/widgets/form.dart (hunk 3 above)
Comment (gitvadim) [reply packages/flutter/lib/src/widgets/form.dart]: My thinking was that if you disable an input, you don't really care about it anymore, so it should not be saved either. I might be wrong, but it seemed to make sense from the consumer point of view. As for doing an assert, is it really the responsibility of the caller to make such assertions? Should the caller really care about the state of the controls instead of the controls making sure they don't do anything they are not expected to?

File: packages/flutter/lib/src/widgets/form.dart (hunk 4 above)
Comment (gitvadim) [reply packages/flutter/lib/src/widgets/form.dart]: Won't the validation error message still appear in the UI unless you clear the state? 
Regarding the autovalidate: I thought about this but then my other though was: why would someone do validation on a disabled input anyway? And I couldn't think of a valid answer.

**Comment by gitvadim** [comment]
Hi,

Let's take this example https://flutter.io/cookbook/forms/validation/. When you have a form with a bunch of FormFields and validators on them and you decide that a certain field is to be disabled, i.e. due to another field having a specific value (i.e. a dropdown) and you attempt to submit the form, the validators will fire regardless of the disabled state of the field. This is quite misleading, because if the disabled field does not pass the validation, you won't be able to submit the form but if it's disabled you probably don't care about it anymore (why would you disable it otherwise?), so there is no reason to trigger the validation in my opinion.

**Comment by Hixie** [comment]
@gitvadim Will you have the opportunity to integrate the above review comments? It's totally fine if you don't have time to do this, we'd be happy to (eventually) take care of it. Thanks for your contribution!

**Comment by gitvadim** [comment]
@Hixie Yes, I was going to do it next week once I get back from my vacation.

File: packages/flutter/lib/src/widgets/form.dart (hunk 4 above)
Comment (gitvadim) [reply packages/flutter/lib/src/widgets/form.dart]: It seems that the input decorator already handles this. At least I did not observe any issues during my manual tests. If you could guide me how to confirm this using unit tests, I'd be glad to write them.

File: packages/flutter/lib/src/widgets/form.dart [hunk 5]
```diff
@@ -256,6 +257,9 @@ class FormField<T> extends StatefulWidget {
   /// autovalidates, this value will be ignored.
   final bool autovalidate;
 
+  /// If true, indicates that the widget is enabled.
+  final bool enabled;
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/form.dart]: docs should be more detailed. What does it mean to be enabled? What does it mean to be disabled? For example, can you still select text when it's disabled? Is validation affected?

File: packages/flutter/lib/src/widgets/form.dart [hunk 6]
```diff
@@ -278,8 +282,9 @@ class FormFieldState<T> extends State<FormField<T>> {
   bool get hasError => _errorText != null;
 
   /// Calls the [FormField]'s onSaved method with the current value.
+  /// If the [FormField]'s enabled is false, the method does nothing.
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/form.dart]: please put a blank line (prefixed with `///`) between the first sentence (which is used as the summary in the docs) and the second (which will be a second paragraph for when you look at it directly).

File: packages/flutter/lib/src/widgets/form.dart [hunk 7]
```diff
@@ -278,8 +282,9 @@ class FormFieldState<T> extends State<FormField<T>> {
...
   void save() {
-    if (widget.onSaved != null)
+    if (widget.onSaved != null && widget.enabled)
       widget.onSaved(value);
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/form.dart]: why do we want this to do nothing when the field is disabled? Is it valid to call this when it's disabled? If so, why, since it does nothing? If not, then we should just assert.

File: packages/flutter/lib/src/widgets/form.dart (hunk 4 above)
Comment (Hixie) [reply packages/flutter/lib/src/widgets/form.dart]: I'm not clear on what you mean by "handles this". What does it do?

The code here IMHO is definitely wrong, for the reasons given above. If we call _validate(), then we should validate. Right now we do the opposite.

Comment (Hixie) [reply packages/flutter/lib/src/widgets/form.dart]: (If we are saying that this should never be called when it's disabled, then we should assert that we are enabled, not clear when we are disabled.)

File: packages/flutter/test/material/text_form_field_test.dart [hunk 8]
```diff
@@ -114,4 +114,28 @@ void main() {
...
+    int _validateCalled = 0;
+
+    await tester.pumpWidget(
+      MaterialApp(
+        home: Material(
+          child: Center(
+            child: TextFormField(
+              enabled: false,
+              autovalidate: true,
+              validator: (String value) { _validateCalled++; return null; },
```
Comment (Hixie) [inline packages/flutter/test/material/text_form_field_test.dart]: nit: prefer `+= 1` for consistency with other code

File: packages/flutter/test/material/text_form_field_test.dart [hunk 9]
```diff
@@ -114,4 +114,28 @@ void main() {
...
+          ),
+        ),
+      ),
+    );
+
+    expect(_validateCalled, 0);
+    await tester.showKeyboard(find.byType(TextField));
+    await tester.enterText(find.byType(TextField), 'a');
+    await tester.pump();
+    expect(_validateCalled, 0);
```
Comment (Hixie) [inline packages/flutter/test/material/text_form_field_test.dart]: we should also check that it does validate if you now enable it.

we also need a lot more tests here, verifying the things we care about (e.g. whether selection works when disabled, whether it renders disabled, whether it's focusable, etc -- basically, all the behaviours you want to make sure we don't regress).

**Comment by jslavitz** [comment]
@gitvadim – Are you going to be able to get to these in the next week or do you want to hand this bug over to us? Absolutely no worries either way!

File: packages/flutter/lib/src/widgets/form.dart (hunk 7 above)
Comment (jslavitz) [reply packages/flutter/lib/src/widgets/form.dart]: It seems pointless because if a field is disabled then the user can't edit it so there's nothing to save. But maybe I'm missing something? Can you think of a case where you'd want to save the values from a disabled field?

File: packages/flutter/lib/src/widgets/form.dart (hunk 5 above)
Comment (gitvadim) [reply packages/flutter/lib/src/widgets/form.dart]: The comment has been expanded to provide more information

File: packages/flutter/lib/src/widgets/form.dart (hunk 6 above)
Comment (gitvadim) [reply packages/flutter/lib/src/widgets/form.dart]: Blank line added

File: packages/flutter/test/material/text_form_field_test.dart (hunk 9 above)
Comment (gitvadim) [reply packages/flutter/test/material/text_form_field_test.dart]: @Hixie, is there a test already where I can see how it can be verified?

**Comment by gitvadim** [comment]
@jslavitz, @Hixie, I've done a few updates to the code based on your comments. I ended up moving the logic of checking whether the widget is disabled or not to the validate() method instead of _validate(). Other than that, I don't see a way how I can implement it. So, should it still be that my solution does not fit the design, I'd be glad if you could take over this change.

**Comment by jslavitz** [comment]
@gitvadim I think that change makes sense, but we still need to add tests, so we're happy to take this PR over from here! Thanks for your contribution!

File: packages/flutter/test/material/text_form_field_test.dart (hunk 9 above)
Comment (Hixie) [reply packages/flutter/test/material/text_form_field_test.dart]: never mind, i misread the patch. no further tests are needed.

**Comment by jslavitz** [comment]
Fixed and merged so I'm closing this!
